biggerquery/workflow.py

```diff
@@ -7,15 +9,164 @@ def __init__(self,
                  schedule_interval=DEFAULT_SCHEDULE_INTERVAL,
                  dt_as_datetime=False,
                  **kwargs):
-        self.definition = definition
+        self.definition = self._parse_definition(definition)
         self.schedule_interval = schedule_interval
         self.dt_as_datetime = dt_as_datetime
         self.kwargs = kwargs
 
     def run(self, runtime):
-        for job in self:
+        for job in self.build_sequential_order():
             job.run(runtime=runtime)
 
-    def __iter__(self):
-        for job in self.definition:
-            yield job
\ No newline at end of file
+    def build_sequential_order(self):
+        return self.definition.sequential_order()
+
+    def call_on_graph_nodes(self, consumer):
+        self.definition.call_on_graph_nodes(consumer)
+
+    def _parse_definition(self, definition):
+        if isinstance(definition, list):
+            return Definition(self._map_to_workflow_jobs(definition))
+        if isinstance(definition, Definition):
+            return definition
+        raise ValueError(f"Invalid argument {definition}")
+
+    @staticmethod
+    def _map_to_workflow_jobs(job_list):
+        return [WorkflowJob(job, i) for i, job in enumerate(job_list)]
+
+
+class WorkflowJob:
+    def __init__(self, job, name):
+        self.job = job
+        self.name = name
+
+    def run(self, runtime):
+        self.job.run(runtime=runtime)
+
+    def __hash__(self):
+        return hash(self.name)
+
+    def __eq__(self, other):
+        return self.name == other.name
+
+    def __str__(self):
+        return "WorkflowJob{job=..., name=%s}" % self.name
+
+
+class Definition:
+    def __init__(self, jobs):
+        self.job_graph = self._build_graph(jobs)
+        self.job_order_resolver = JobOrderResolver(self.job_graph)
+
+    def sequential_order(self):
+        return self.job_order_resolver.find_sequential_run_order()
+
+    def call_on_graph_nodes(self, consumer):
+        return self.job_order_resolver.call_on_graph_nodes(consumer)
+
+    def _build_graph(self, jobs):
+        if isinstance(jobs, list):
+            job_graph = self._convert_list_to_graph(jobs)
+        elif isinstance(jobs, dict):
+            job_graph = jobs
+        else:
+            raise ValueError("Job graph has to be dict or list")
+
+        JobGraphValidator(job_graph).validate()
+        return job_graph
+
+    @staticmethod
+    def _convert_list_to_graph(job_list):
+        graph_as_dict = OrderedDict()
+        if len(job_list) == 1:
+            graph_as_dict[job_list[0]] = []
+        else:
+            for i in range(1, len(job_list)):
+                graph_as_dict[job_list[i - 1]] = [job_list[i]]
```

Comment:
What's going on here?

Reply:
It is the part of the code necessary to preserve backward compatibility. Currently we define workflows by passing list of jobs except graph. New algorithm for running jobs (or converting them to airflow dag) operate on graphs internally. To be able to adjust the old interface to the new algorithm we have to convert list of jobs to graph.